Add store test case for GeoRadiusByMember

azuredream opened the pull request:
#1548

azuredream commented:
Please refer to PR2149 for the history.
I'm still seeing the same issue, store worked in CLI but not in test:

I encountered a wired issue. This command works in CLI:
127.0.0.1:6379> GEORADIUSBYMEMBER Europe Madrid 700 KM STORE test1
(integer) 2
(15.54s)
127.0.0.1:6379> zrange test1 0 -1
1) "Madrid"
2) "Lisbon"

The first unit test case worked as well.
`GEORADIUSBYMEMBER Europe Madrid 700 KM WITHCOORD WITHDIST`

However, **the second test case failed. Please note that this command worked in CLI.**
`GEORADIUSBYMEMBER Europe Madrid 700 KM STORE store_key`
[ RUN      ] ZSetFamilyTest.GeoRadiusByMember
get resp: ERR Member not found
F20231114 21:37:55.006270 2607605 test_utils.cc:466] Check failed: RespExpr::STRING == int(resp.type) (0 vs. 6) [GEORADIUSBYMEMBER, Europe, Madrid, 700, KM, STORE, store_key]

When I used GDB to run the test case line by line, it worked again.

I suppose it's something to do with the schedule-execute model.
When I use `STORE` option, there will be 3 transactions while 2 without STORE.
It's confusing to me that the test case failed at the first transaction (get key from db), not the last one.
I'd appreciate it if you could share your thought on this. Thanks!

romange commented on src/server/zset_family.cc:
```diff
@@ -2733,8 +2737,14 @@ void ZSetFamily::GeoDist(CmdArgList args, ConnectionContext* cntx) {
+//            False: Using transaction->Execute(,True)
 bool MembersOfAllNeighbors(ConnectionContext* cntx, string_view key, const GeoHashRadius& n,
-                           const GeoShape& shape_ref, GeoArray* ga, unsigned long limit) {
+                           const GeoShape& shape_ref, GeoArray* ga, unsigned long limit,
```
can it be that you call `MembersOfAllNeighbors(...,, true, false)` ?

romange commented on src/server/zset_family.cc:
```diff
@@ -3028,16 +3079,19 @@ void ZSetFamily::GeoRadiusByMember(CmdArgList args, ConnectionContext* cntx) {
-  if (result.status() == OpStatus::WRONG_TYPE) {
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute(std::move(cb), false);
+  if (member_score.status() == OpStatus::WRONG_TYPE) {
```
you can not call `Execute(..)` and then exit early. In fact, please a test that runs this command on a wrong type and see the testing framework fails because of this early return. 

If you want to return early you should call `transaction->Conclude();`

azuredream commented:
Hi romange, I couldn't make `STORE` and `STOREDIST`` work shortly. Schedule-execute and sharding are confusing to me. I'll have to learn more about the source code. A guess is it that I should have used `ScheduleSingleHopT` instead of `schedule-execute`. But  more issue came out. Maybe someone else knows how to write DB correctly. 

Just updated my PR to add test cases and fixed a bug in GeoSearch.

romange reviewed (approved):
Thank you for your contribution!

romange commented:
@azuredream I see you reverted the parsing code as well, is it intended?

azuredream commented:
I just put the parsing code back. Thanks for the reminder
